src/tests/Common/scripts/nativeaottest.cmd

```diff
@@ -21,6 +21,10 @@ if not exist %__JsFilePath% (
   set __JsFilePath=%__JsFilePath:~0,-3%.mjs
 )
 
+if not exist %__JsFilePath% (
+  set __JsFilePath=%1\native\main.js
+)
```

Comment:
It is non standard for .NET projects. On web it's up to the user how he names his entrypoint. Nothing settled IMO. Our templates has main.js as entrypoint.

I can rename main.js in the test to DotnetJs.js (as the project itself) if you prefer that

Reply:
`bike_shed.js` 😉

Reply:
No, it's fine to leave this as it is for now. I suppose it is not really the `main` file I am worried about, but the all of `dotnet.*` files. Ideally, the target there would be to have the same output as what Emscripten generates: one `Project.js`, one `Project.wasm`. It helps optics/perception. Nothing to do about that at this point, though.

Reply:
We have multiple files for performance and modularity reasons. We are not going back to single JS file. It's more trouble than worth to live inside emscripten's JS postprocessing tooling. Also you need extra JS file for threads.

Reply:
Actually I think we will have more modules, so that we could load some of the larger JS components only when necessary. Download size and startup speed are tier 1) goal for us.

Reply:
We could discuss more here https://github.com/dotnet/runtime/issues/86162

Reply:
No, I don't think doing most it in C# is possible. But I'm open to see specific improvements.

Reply:
Ideally on main branch.

Reply:
> No, I don't think doing most it in C# is possible. But I'm open to see specific improvements.

I think I was bit blunt here, let me explain.
- big part of the JS/TS we have deals with download of assets and startup. That can't be C#, chicken/egg. It's heavily async, using network badwidth and CPU at the same time.
- the implementation of the JS interop. chicken/egg.
- jiterpreter, also kind of chicken/egg. Maybe this could become C. But TS is easier to maintain.
- whatever emscripten/POSIX is not good enough for us. Hacks into emscripten JS internals, small code size
- timers, single-thread "thread pool", `Task` asynchrony all of which has to play well with browser event loop.
- fancy trade-offs, like hybrid globalization, using browser's APIs instead of ICU. There is very brittle balance of perf vs size. How you pass buffers makes huge difference. Maybe this could be a candidate for moving it to low level C# with raw pointers etc. [Example ](https://github.com/dotnet/runtime/pull/95583)of how we are working on good tradeoff right now.
- http and WS clients. There is thin TS code that wraps the browser API in a way that makes it possible to consume as JS interop. Dealing with buffers and lack cancellation tokens in JS.

Well, but perhaps you could bring fresh perspective @SingleAccretion :)

Reply:
jiterpreter is something that could become separate ES6 module, or mono C code or RollupJS conditional compilation. Because it only takes bytes for NativeAOT right now.

Reply:
JS is well trimmable, but we don't have things like rollupjs in the workload/emscripten bundle. So we don't have the tools on customer machine, where he sets the feature flags. Maybe one they we will add rollupjs to the workload and shake the tree some more. 

emscripten makes JS trimmable even together with C code, but it only works if you are willing to write your JS in C style. (no closures). It also damages your code if you use modern ES syntax. We are not willing to try again. If you want to have look, see Net6 codebase we had.

Reply:
> My perspective is "pay for play" design. 

We also prefer that.

> (e. g.: only link in JS `string` marshalling if something has a JSImport/JSExport with a `string` in it)

linker across C#/C/JS is though. Especially with JS being very dynamic. 

In this specific example, we interpret the JS side of the "marshal single argument" from metadata about the method signature. And so we have to have all marshallers available on JS side. 

Alternative is to ship generated JS code for each JSImport/JSExport. It has tooling and deployment troubles. Roslyn doesn't support that either.

Reply:
> We are not willing to try again. If you want to have look, see Net6 codebase we had.

Just curious - was it 'simply' that the code was ugly, or there was some more fundamental problem (too many Emscripten bugs, hard to track down failure modes, ...)?

> In this specific example, we interpret the JS side of the "marshal single argument" from metadata about the method signature. And so we have to have all marshallers available on JS side.

Its possible to write it in such a way so as to make the dependencies explicit. For example, you can imagine that for:
```
[JSImport] void func(string);
```
The generator would generate:
```
[DllImport("*")] void js_marshal_string();
[JSImport, DependsOn("js_marshal_string")] void func(string); // Imaginary attribute; just for the sake of argument
```
That imaginary attribute would lead to passing `--undefined:js_marshal_string` to the linker, which would preserve that marshalling method in a C-like static linking scenario. Admittedly, this is questionable design, but good enough for illustrative purposes.

It is fine to say that "well, this is a rather complicated thing to do, so we will settle for some middle-of-the-road design, like making trimming based in feature switches an JS module (one module per switch)", but then you really have to be careful with the amount of JS (hence my earlier point about avoiding it).

Reply:
> Alternative is to ship generated JS code for each JSImport/JSExport. It has tooling and deployment troubles. Roslyn doesn't support that either.

I'm dreaming about that one ❤️

Reply:
> Just curious - was it 'simply' that the code was ugly, or there was some more fundamental problem (too many Emscripten bugs, hard to track down failure modes, ...)?

Just go and have look :)

> [DllImport("*")] void js_marshal_string();
> [JSImport, DependsOn("js_marshal_string")] void func(string); // Imaginary attribute; just for the sake of argument

And to use that to fabricate roots for JS trimmer, right ?
`JSExport` could be more difficult.
Would that be multi-pass ?

> It is fine to say that "well, this is a rather complicated thing to do, so we will settle for some middle-of-the-road design, like making trimming based in feature switches and JS modules (one module per switch)", but then you really have to be careful with the amount of JS 

That's exactly where we are.

> (hence my earlier point about avoiding it).

In that sense I agree, but not as top priority or goal on it's own.